File: scm-java/src/main/scheme/primitives/PrimitiveCreateModule.java
package scheme.primitives;

import scheme.Modules;
import scheme.Primitive;
import scheme.SourcePos;

public class PrimitiveCreateModule extends Primitive
{
    private Modules modules;

    public PrimitiveCreateModule(Modules modules) {
        this.modules = modules;
    }

    @Override
    public String name() {
        return "%create-module";
    }

    @Override
    public String info() {
        return "Syntax: (%create-module module-name)\n" +
               "Library: (scm core)\n" +
               "Description: Creates the named module if it does not already exist. Returns the module name.\n" +
               "Example:\n" +
               "  (create-module '(my lib))";
    }

    @Override
    public Object apply(SourcePos pos, Object[] arguments) {
        checkArgs(pos, arguments, 1, 1);

        var oldModule = modules.getCurrentModule();
        modules.setCurrentModule(Modules.asModuleName(arguments[0]));
        modules.setCurrentModule(oldModule.getName());
        return arguments[0];
    }
}

File: scm-csharp/scheme/primitives/PrimitiveCreateModule.cs
namespace scheme;

public class PrimitiveCreateModule : Primitive
{
    private Modules modules;

    public PrimitiveCreateModule(Modules modules)
    {
        this.modules = modules;
    }

    public override string Name() => "%create-module";

    public override string Info() =>
        "Syntax: (%create-module module-name)\n" +
        "Library: (scm core)\n" +
        "Description: Creates the named module if it does not already exist. Returns the module name.\n" +
        "Example:\n" +
        "  (create-module '(my lib))";

    public override object Apply(SourcePos? pos, object[] arguments)
    {
        CheckArgs(pos, arguments, 1, 1);
        var oldModule = modules.GetCurrentModule();
        modules.SetCurrentModule(Modules.AsModuleName(arguments[0]));
        modules.SetCurrentModule(oldModule.Name);
        return arguments[0];
    }
}
